llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp

```diff
@@ -655,25 +655,31 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
         }
       }
 
-      // TODO: If we only want bits that already match the signbit then we don't
+      // We only want bits that already match the signbit then we don't
       // need to shift.
+      uint64_t ShiftAmt = SA->getLimitedValue(BitWidth - 1);
+      if (DemandedMask.countr_zero() >= ShiftAmt) {
+        unsigned NumHiDemandedBits = BitWidth - DemandedMask.countr_zero();
+        unsigned SignBits =
+            ComputeNumSignBits(I->getOperand(0), Depth + 1, CxtI);
```

Comment:
Can we just check `nsw` here? We will set `nsw` if `computeNumSignBits >= ShCnt` + other conditions.

Reply:
It seems that knowing that `computeNumSignBits` is greater than `ShCnt` is not enough. We need to shift `computeNumSignBits` by `ShCnt` and then check whether `DemandedBits` is still within the range of `SignBits`. 
If there is a way to check the range of `SignBits` and `DemandedBits` by checking `nsw`, please let me know.

Reply:
Ah you are right, although you should be able to check `nsw` first before doing the `computeNumSignBits` as w.o `nsw` think we always know `ShCnt > computeNumSignBits`.

Reply:
@goldsteinn I understand!